hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -415,68 +415,74 @@ private class Emit(
                   codeCnsq.setup,
                   codeAltr.setup))),
             const(false),
-            Code._empty)
+            PValue._empty)
         } else {
-          val typ = ir.typ
           val codeCond = emit(cond)
-          val out = coerce[Any](mb.newLocal()(typeToTypeInfo(typ)))
+          val out = mb.newPLocal(pt)
           val mout = mb.newLocal[Boolean]()
           val codeCnsq = emit(cnsq)
           val codeAltr = emit(altr)
 
           val setup = Code(
             codeCond.setup,
             codeCond.m.mux(
-              Code(mout := true, out := defaultValue(typ)),
+              Code(mout := true, out := pt.defaultValue),
               coerce[Boolean](codeCond.v).mux(
-                Code(codeCnsq.setup, mout := codeCnsq.m, out := mout.mux(defaultValue(typ), ir.pType.copyFromTypeAndStackValue(mb, er.region, cnsq.pType, codeCnsq.v))),
-                Code(codeAltr.setup, mout := codeAltr.m, out := mout.mux(defaultValue(typ), ir.pType.copyFromTypeAndStackValue(mb, er.region, altr.pType, codeAltr.v))))))
-
-          EmitTriplet(setup, mout, out)
+                Code(codeCnsq.setup,
+                  mout := codeCnsq.m,
+                  mout.mux(
+                    out := pt.defaultValue,
+                    out := ir.pType.copyFromPValue(mb, er.region, codeCnsq.pv))),
+                Code(codeAltr.setup,
+                  mout := codeAltr.m,
+                  mout.mux(
+                    out := pt.defaultValue,
+                    out := ir.pType.copyFromPValue(mb, er.region, codeAltr.pv))))))
+
+          EmitTriplet(setup, mout, out.load())
         }
 
       case Let(name, value, body) =>
-        val vti = typeToTypeInfo(value.typ)
         val mx = mb.newField[Boolean]()
-        val x = coerce[Any](mb.newField(name)(vti))
+        val x = mb.newPField(name, value.pType)
         val storeV = wrapToMethod(FastIndexedSeq(value)) { (_, _, codeV) =>
           Code(codeV.setup,
             mx := codeV.m,
-            x := mx.mux(defaultValue(value.typ), codeV.v))
+            mx.mux(
+              x := value.pType.defaultValue,
+              x := codeV.pv))
         }
-        val bodyenv = env.bind(name, (vti, mx.load(), x.load()))
+        val bodyenv = env.bind(name, (mx.load(), x.load()))
         val codeBody = emit(body, env = bodyenv)
         val setup = Code(
           storeV,
           codeBody.setup)
 
-        EmitTriplet(setup, codeBody.m, codeBody.v)
-      case Ref(name, typ) =>
-        val ti = typeToTypeInfo(typ)
-        val (t, m, v) = env.lookup(name)
-        assert(t == ti, s"$name type annotation, $typ, $t doesn't match typeinfo: $ti")
+        EmitTriplet(setup, codeBody.m, codeBody.pv)
+      case Ref(name, _) =>
+        val (m, v) = env.lookup(name)
+        // FIXME Why is the strip necessary?
```

Comment:
Maybe better as an issue?

Reply:
OK, I fixed this.  The ptypes must now agree exactly (as they should).  There might be some lingering failures.